### ISSUE TITLE
Bugs and more in FS and/or FFat and SPIFFS on ESP32

### ISSUE THREAD

**Issue by ullix** [espressif/arduino-esp32#3927]
I used FFat on ESP32 very happily, but now ran into some serious issues.

I want to create a (large) file, and write fixed-size records at varying positions, actually cycling through the file and thereby overwriting old records. This does not work, because FFat allows me to seek to a certain position (and read from it), but it does NOT allow me, to write at this position. This looks like a bug to me, or is it a feature?

In file FS.h ( lib prop: name=FS, version=1.0, author=Hristo Gochkov, Ivan GrokhtkovI ) find these definitions:

        #define FILE_READ       "r"
        #define FILE_WRITE      "w"      
        #define FILE_APPEND     "a" 

so I guess, "r, w, a" is all, what is available to me.

I reduced my sketch down to this simple one (error checking omitted):

        #include "FS.h"
        #include "FFat.h"

        String      testfile = "/test.log";
        uint16_t    counter  = 10;

        void printFileContent(){
            Serial.println("----------Print File Content:-----------------------");

            File file = FFat.open(testfile, "r");
            while(file.available()){
                Serial.write(file.read());
            }
            file.close();
            Serial.println();
        }

        void writeToTestfile(int index, String msg){
            Serial.println("----------Write to File, then Print File Content:-------------------");
            File file = FFat.open(testfile, "rw");
            file.seek(index);

            Serial.printf("writeToTestfile: before file.write seekpos:%i, fposition:%i, fsize:%i, Strlen:%i, String:%s \n", index, file.position(), file.size(), msg.length(), msg.c_str());
            file.write((const uint8_t *) msg.c_str(), msg.length());
            //~ file.print(msg);
            Serial.printf("writeToTestfile: after  file.write seekpos:%i, fposition:%i, fsize:%i, Strlen:%i, String:%s \n", index, file.position(), file.size(), msg.length(),  msg.c_str());

            file.flush();
            Serial.printf("writeToTestfile: after  file.flush seekpos:%i, fposition:%i, fsize:%i, Strlen:%i, String:%s \n", index, file.position(), file.size(), msg.length(),  msg.c_str());

            file.close();
        }

        void setup() {
            Serial.begin(115200);
            Serial.setDebugOutput(true);

            FFat.begin(false, "/ffat", 10);

            File file = FFat.open(testfile, FILE_WRITE);
            for (int i=0; i < 256; i++){
                file.print('i');
            }
            file.close();
        }

        void loop(){

            if(Serial.available()) {
                String command = Serial.readStringUntil('\n');
                if(command.startsWith("p")){
                    printFileContent();
                    Serial.println();
                }
                else if (command.startsWith("w")){
                    writeToTestfile(counter, "test");
                    counter += 20;
                    if (counter > 1000) counter = 1;
                    printFileContent();
                    Serial.println();
                }
            }
            delay(100);

        }

In setup() I create a 256 bytes file, and fill it complete with 'i's.

In loop() I check the terminal and on 'p' input I print out the complete file, and on 'w' I write the text 'test' to increasing positions in the file, and then print out the file.

With the testfile opened for 'rw' (line 25) the seek works, the position is correct, but nothing at all is written, not at position 10 nor anywhere else:

        ----------Print File Content:-----------------------
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii

        ----------Write to File, then Print File Content:-------------------
        writeToTestfile: before file.write seekpos:10, fposition:10, fsize:256, Strlen:4, String:test
        writeToTestfile: after  file.write seekpos:10, fposition:10, fsize:256, Strlen:4, String:test
        writeToTestfile: after  file.flush seekpos:10, fposition:10, fsize:256, Strlen:4, String:test
        ----------Print File Content:-----------------------
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii

OK, perhaps you can't combine to 'rw', so I repeat with 'w' only:

        ----------Print File Content:-----------------------
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii

        ----------Write to File, then Print File Content:-------------------
        writeToTestfile: before file.write seekpos:10, fposition:10, fsize:256, Strlen:4, String:test
        writeToTestfile: after  file.write seekpos:10, fposition:14, fsize:0, Strlen:4, String:test
        writeToTestfile: after  file.flush seekpos:10, fposition:14, fsize:0, Strlen:4, String:test
        ----------Print File Content:-----------------------
        iiiiiiiiiitest

        ----------Write to File, then Print File Content:-------------------
        writeToTestfile: before file.write seekpos:30, fposition:30, fsize:14, Strlen:4, String:test
        writeToTestfile: after  file.write seekpos:30, fposition:34, fsize:0, Strlen:4, String:test
        writeToTestfile: after  file.flush seekpos:30, fposition:34, fsize:0, Strlen:4, String:test
        ----------Print File Content:-----------------------
        iiiiiiiiiitestiiiiiiiiiiiiiiiitest

So now it positions at 10, writes 4 bytes, and finds itself at position 14. Ok. But now it has truncated the file after writing - not what I wanted. And the filesize is reported as 0 (zero)! That is clearly a bug.

Re-issuing the command 'w' does the write again, and sees the filesize at 14. Correct. Yet, it has no problem to position at 30, writes 'test', and find a zero filesize. Positioning outside the file without an error? another bug. (In doing so it found the 'i's it had cutoff before).

Last option, open the file with 'a', append.:

        ----------Print File Content:-----------------------
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii

        ----------Write to File, then Print File Content:-------------------
        writeToTestfile: before file.write seekpos:10, fposition:10, fsize:256, Strlen:4, String:test
        writeToTestfile: after  file.write seekpos:10, fposition:260, fsize:256, Strlen:4, String:test
        writeToTestfile: after  file.flush seekpos:10, fposition:260, fsize:256, Strlen:4, String:test
        ----------Print File Content:-----------------------
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiitest

        ----------Write to File, then Print File Content:-------------------
        writeToTestfile: before file.write seekpos:30, fposition:30, fsize:260, Strlen:4, String:test
        writeToTestfile: after  file.write seekpos:30, fposition:264, fsize:260, Strlen:4, String:test
        writeToTestfile: after  file.flush seekpos:30, fposition:264, fsize:260, Strlen:4, String:test
        ----------Print File Content:-----------------------
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiitesttest

It does the positioning, but then appends the 'test' to the end of the file, the normal appending behavior. It then finds itself at position 260, yet it reports a file size of 256 only, even after flushing? Another bug, I believe.

(I tried the 'r' opening option for completeness also: it behaved exactly as the first tested 'rw' option).

Apart from all the bugs: is it really impossible to write something in the middle of an existing file? 

I FFat does not allow this - is there any filesystem out, which does? Or is not FFat but FS the real culprit? Which would mean that there is no filesystem out there, which doies allow this? Hard to believe.

**Comment by ullix** [comment]
I have a terrible experience with SPIFFS, as I had reported here: [https://forum.arduino.cc/index.php?topic=648987.new](url), but wanted to give it a shot.

In my code above I changed only "FFat" against "SPIFFS" (and changed the ESP partitioning and formatted once with SPIFFS.format();), and then tested the same options:

"rw" : exactly the same as with FFat; nothing gets written at all
"w" : it seeks pos 10, but actually uses pos 0, then writes 'test' and truncates the rest. But it gets the new file size= 4 right
"a" : goes to position 10, but writes at position 256. But gets the file size right with 260.

Ughh.

**Comment by lbernstone** [comment]
TL;DR
The file write operations do not work like a word processor.  If you want to insert into the middle of a file, you need to capture all the trailing data in a char array, seek to the insert point in the file (opened for write), add your data, then append the captured data.

**Comment by ullix** [comment]
> TL;DR

I can see that. Because I never said anything about **inserting** anything, and thereby increasing the size of the file, but my 2nd sentence already said **"overwriting old records"**.

This is a meaningful operation for cyclic logs with fixed-size records: after the last record in a fixed size file had been reached, the first record will be overwritten. You will loose old records, but have a constant number of records.

But - unless I missed an option - this does not seem possible, neither on FFat nor on SPIFFS? And on top of this, there are a few bugs.

**Comment by lbernstone** [comment]
Flush does have a bug, which appears to be upstream.  The following modification (in setup) of FFat_Test.ino works for me:
```
    listDir(FFat, "/", 0);
    writeFile(FFat, "/hello.txt", "This is a test file 0123456789 ");
    File f = FFat.open("/hello.txt", "w");
    f.seek(22);
    f.print("added ");
    f.close();
    appendFile(FFat, "/hello.txt", "World!\r\n");


Writing file: /hello.txt
- file written
Appending to file: /hello.txt
- message appended
Reading file: /hello.txt
- read from file:
This is a test file 01added World!
```

**Comment by ullix** [comment]
Sure, it works. You are pointing out the very problem I have been showing above. You are positioning to 22, writing "added ", thereby cutting off "789 ", and then append "World". (In SPIFFS even that would not have worked).

Now do it this way: imagine your flash is completely used up, no space left to save anything, not in flash, not in RAM. Now in your file "hello.txt" replace the word "test" with the word "TEST". Can you do that?

**Comment by ullix** [comment]
Yeehaa, isn't it wonderful when people adhere to standards, like [http://www.cplusplus.com/reference/cstdio/fopen/](url) (at least partially), and all you have to do is to read it?

The "rb+" option (see line 25 in code: File file = FFat.open(testfile, "rb+");) did the trick:

        ----------Print File Content:-----------------------
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii

        ----------Write to File, then Print File Content:-------------------
        writeToTestfile: before file.write seekpos:10, fposition:10, fsize:256, Strlen:4, String:test
        writeToTestfile: after  file.write seekpos:10, fposition:14, fsize:256, Strlen:4, String:test
        writeToTestfile: after  file.flush seekpos:10, fposition:14, fsize:256, Strlen:4, String:test
        ----------Print File Content:-----------------------
        iiiiiiiiiitestiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii
        iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii

I tried the "wb+" and "ab+" options also for completeness, and the result was the same as before, in agreement with the standard.

Not sure when the "b" must or must not be used, but I will be dealing with binary files anyway, so I'll use it.

Well, that part is ok now, but the bugs remain, like reporting wrong file size, or allowing a positioning outside the file.
